sql/2021/resource-hints/hints_adoption_by_rank.sql

```diff
@@ -0,0 +1,61 @@
+#standardSQL
+# % of sites that use each type of resource hint grouped by rank
+
+CREATE TEMPORARY FUNCTION getResourceHints(payload STRING)
+RETURNS STRUCT<preload BOOLEAN, prefetch BOOLEAN, preconnect BOOLEAN, prerender BOOLEAN, `dns-prefetch` BOOLEAN, `modulepreload` BOOLEAN>
+LANGUAGE js AS '''
+var hints = ['preload', 'prefetch', 'preconnect', 'prerender', 'dns-prefetch', 'modulepreload'];
+try {
+  var $ = JSON.parse(payload);
+  var almanac = JSON.parse($._almanac);
+  return hints.reduce((results, hint) => {
+    results[hint] = !!almanac['link-nodes'].nodes.find(link => link.rel.toLowerCase() == hint);
+    return results;
+  }, {});
+} catch (e) {
+  return hints.reduce((results, hint) => {
+    results[hint] = false;
+    return results;
+  }, {});
+}
+''' ;
+
+SELECT
+  client,
+  COUNTIF(hints.preload) AS preload,
+  COUNTIF(hints.preload) / COUNT(0) AS pct_preload,
+  COUNTIF(hints.prefetch) AS prefetch,
+  COUNTIF(hints.prefetch) / COUNT(0) AS pct_prefetch,
+  COUNTIF(hints.preconnect) AS preconnect,
+  COUNTIF(hints.preconnect) / COUNT(0) AS pct_preconnect,
+  COUNTIF(hints.prerender) AS prerender,
+  COUNTIF(hints.prerender) / COUNT(0) AS pct_prerender,
+  COUNTIF(hints.`dns-prefetch`) AS dns_prefetch,
+  COUNTIF(hints.`dns-prefetch`) / COUNT(0) AS pct_dns_prefetch,
+  COUNTIF(hints.modulepreload) AS modulepreload,
+  COUNTIF(hints.modulepreload) / COUNT(0) AS pct_modulepreload,
+  rank
+FROM (
+  SELECT
+    _TABLE_SUFFIX AS client,
+    url AS page,
+    getResourceHints(payload) AS hints
+  FROM
+    `httparchive.pages.2021_07_01_*`
+)
+JOIN (
+    SELECT
+      _TABLE_SUFFIX AS client,
+      url AS page,
+      rank
+    FROM
+      `httparchive.summary_pages.2021_07_01_*`
+)
+USING
+  (client, page)
```

Comment:
By default ranks are exclusive, so for example the top 100,000 rank contains 99,000 URLs because the top 1,000 is its own exclusive segment. So to ensure that the top 100k contains 100k URLs we combine them like this:
```suggestion
  (client, page),
  UNNEST([1000, 10000, 100000, 1000000, 10000000]) AS rank
WHERE
  _rank <= rank
```

Reply:
Good to know. Would suggest having this documented somewhere - unless it already is and I missed it.

Reply:
Yeah should be in Analysts guide. Will add...